ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/State.hs

```diff
@@ -580,32 +274,39 @@ fromList k n (anchor, signers, ebbs) =
 serializationFormatVersion0 :: VersionNumber
 serializationFormatVersion0 = 0
 
-encodePBftState :: Serialise (PBftVerKeyHash c)
-                => PBftState c -> Encoding
+compact :: PBftCrypto c => PBftState c -> Map (PBftVerKeyHash c) [SlotNo]
+compact =
+      Foldable.foldl'
+        (\acc (PBftSigner slot key) -> Map.insertWith (<>) key [slot] acc)
+        Map.empty
+    . inWindow
+
+decompact :: PBftCrypto c => Map (PBftVerKeyHash c) [SlotNo] -> PBftState c
+decompact =
+      fromList
+    . sortOn pbftSignerSlotNo
+      -- We use a difference list to avoid repeated (7 lists in total)
```

Comment:
Minor: maybe just use `concatMap`, since the outer list is always just 7?

Reply:
Yeah, you're right, premature optimisation, etc. Each list of the 7 lists will have 2160/7 elements. This is not a known hotspot, so simplicity trumps performance.